aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp

```diff
@@ -89,52 +92,136 @@ m_getObjectRetries(0),
 m_listObjectsRetries(0),
 m_headBucketRetries(0)
 {
-    if (m_fileStream.good() && m_fileStream.is_open())
+    if (useStream)
     {
-        SetFileSize(static_cast<uint64_t>(m_fileStream.tellg()));
+        m_inputStream = fileStream;
+    }
+    else
+    {
+        m_inputStream = Aws::MakeShared<Aws::FStream>(ALLOCATION_TAG, fileName.c_str(), std::ios::in | std::ios::binary | std::ios::ate);
+    }
+    
+    if (m_inputStream->good())
+    {
+        // If input stream is a file stream, make sure it's open
+        if (auto fileStream = std::dynamic_pointer_cast<Aws::FStream>(m_inputStream))
+        {
+            if (!fileStream->is_open())
+            {
+                CompletionFailure("Failed to open file.");
+                return;
+            }
+        }
+        
+        m_inputStream->seekg(0, std::ios_base::end);
+        SetFileSize(static_cast<uint64_t>(m_inputStream->tellg()));
         m_bytesRemaining = GetFileSize();
-        m_fileStream.seekg(0);
+        m_inputStream->seekg(0);
     }
     else
     {
         CompletionFailure("Failed to open file.");
         return;
     }
-   
+    
     if (GetFileSize())
     {
         m_totalParts = 1 + static_cast<uint32_t>((GetFileSize() - 1) / MB5_BUFFER_SIZE); // How many total buffer operations are we performing
     }
+    else
+    {
+        m_totalParts = 1;
+        m_isEmptyFile = true;
+    }
 }
-
+    
 UploadFileRequest::UploadFileRequest(const Aws::String& fileName,
                                      const Aws::String& bucketName,
                                      const Aws::String& keyName,
                                      const Aws::String& contentType,
+                                     Aws::Map<Aws::String, Aws::String>&& metadata,
                                      const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
                                      bool createBucket,
                                      bool doConsistencyChecks) :
-UploadFileRequest(fileName, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>{}, s3Client, createBucket, doConsistencyChecks)
+UploadFileRequest(fileName, nullptr, bucketName, keyName, contentType, std::move(metadata), s3Client, executor, createBucket, doConsistencyChecks, false)

```

Comment:
I'm not sure why I have to use std::move here in order to call the other constructor. Since `metadata` is an r-value, shouldn't it be possible to pass it directly to the other constructor (which also takes a `Aws::Map<Aws::String, Aws::String>&&`)?